### PR TITLE
Simplify the internal API for declaring command-line options

### PR DESCRIPTION
**Pull request by Zalathar**
The internal APIs for declaring command-line options are old, and intimidatingly complex. This PR replaces them with a single function that takes explicit `stability` and `kind` arguments, making it easier to see how each option is handled, and whether it is treated as stable or unstable.

We also don't appear to have any tests for the output of `rustc --help` and similar, so I've added a run-make test to verify that this PR doesn't change any output. (There is already a similar run-make test for rustdoc's help output.)

---

The librustdoc changes are simply adjusting to updated compiler APIs; no functional change intended.

---

A side-effect of these changes is that rustfmt can once again format the entirety of these option declaration lists, which it was not doing before.

### CODE REVIEW

**Comment by Zalathar** [comment]
The librustdoc changes are simply adjusting to updated compiler APIs; no functional change intended.

File: compiler/rustc_session/src/config.rs [hunk 1]
```diff
@@ -1464,46 +1458,60 @@ The default is {DEFAULT_EDITION} and the latest stable edition is {LATEST_STABLE
...
-                             static, framework, or dylib (the default).
-                             Optional comma separated MODIFIERS (bundle|verbatim|whole-archive|as-needed)
-                             may be specified each with a prefix of either '+' to
-                             enable or '-' to disable.",
+            "Link the generated crate(s) to the specified native\n\
+                library NAME. The optional KIND can be one of\n\
+                static, framework, or dylib (the default).\n\
+                Optional comma separated MODIFIERS\n\
+                (bundle|verbatim|whole-archive|as-needed)\n\
+                may be specified each with a prefix of either '+' to\n\
```
Comment (Zalathar) [inline compiler/rustc_session/src/config.rs:1497]: IIRC this is one of the entries that used to confuse rustfmt, and cause the rest of the list to quietly not be auto-formatted at all.

Replacing the hard line-breaks with `\n\` seems to work around that, without changing the output seen by the new run-make test.

File: tests/run-make/rustc-help/help-v.diff [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
++        --remap-path-prefix FROM=TO
++                        Remap source names in all output (compiler messages
++                        and output files)
++    @path               Read newline separated options from `path`
+ 
+ Additional help:
+     -C help             Print codegen options
+     -W help             Print 'lint' options and default settings
+     -Z help             Print unstable compiler options
+-    --help -v           Print the full set of options rustc accepts
```
Comment (GuillaumeGomez) [inline tests/run-make/rustc-help/help-v.diff:28]: Why was this removed?

Comment (Zalathar) [reply tests/run-make/rustc-help/help-v.diff:28]: This isn't a removed line; it's a diff file showing the difference between the output of `rustc --help` and `rustc --help -v`.

Comment (Zalathar) [reply tests/run-make/rustc-help/help-v.diff:28]: (This is existing behaviour; passing `--help -v` removes the message telling you that you can pass `--help -v`.)

**Review by GuillaumeGomez — APPROVED** [review]
Changes look good to me, thanks! I'll let someone from the compiler team confirm it's good on their side as well.

**Comment by Zalathar** [comment]
Hmm, it occurs to me that the help output is slightly different on beta/stable, but only if `RUSTC_BOOTSTRAP` is not set.

Looking at some of the other run-make tests, it seems that they already pass `-Z` flags to the compiler without any special nightly-only directives, so hopefully that variable is set and things will be fine? Even if it somehow isn't fine, I guess the worst-case outcome is that someone is slightly inconvenienced when doing the next beta branch.

**Comment by jieyouxu** [comment]
> Hmm, it occurs to me that the help output is slightly different on beta/stable, but only if `RUSTC_BOOTSTRAP` is not set.
> 
> Looking at some of the other run-make tests, it seems that they already pass `-Z` flags to the compiler without any special nightly-only directives, so hopefully that variable is set and things will be fine? Even if it somehow isn't fine, I guess the worst-case outcome is that someone is slightly inconvenienced when doing the next beta branch.

`compiletest` is run with `RUSTC_BOOTSTRAP=1`

https://github.com/rust-lang/rust/blob/4b198d6871e4a3826f01f7d6111b1c3484853cd8/src/bootstrap/src/core/build_steps/test.rs#L2091

File: compiler/rustc_session/src/config.rs [hunk 3]
```diff
@@ -1468,46 +1472,60 @@ The default is {DEFAULT_EDITION} and the latest stable edition is {LATEST_STABLE
...
+        opt(Stable, Flag, "h", "help", "Display this message", ""),
+        opt(
+            Stable,
+            Multi,
+            "",
+            "cfg",
+            "Configure the compilation environment.\n\
+                SPEC supports the syntax `NAME[=\"VALUE\"]`.",
+            "SPEC",
+        ),
```
Comment (jieyouxu) [inline compiler/rustc_session/src/config.rs:1476]: Question: what does `rustc_short_optgroups` actually mean?

> Returns the "short" subset of the rustc command line options

Like the options that have a shorthand...?

**Review by jieyouxu — APPROVED** [review]
Thank you, this looks like a significant readibility cleanup for the options handling! One minor question, but otherwise looks good.

File: compiler/rustc_session/src/config.rs (hunk 3 above)
Comment (jieyouxu) [reply compiler/rustc_session/src/config.rs:1476]: Anyway, this doesn't have to be in this PR (probably for the better).

Comment (Zalathar) [reply compiler/rustc_session/src/config.rs:1476]: I believe it means “the subset of options that we print when running `rustc --help` without `-v`”.

It’s not a great name!
